File: compiler/bench/luhn.c
// expected: 200000
// Luhn checksum benchmark - validate credit card numbers, run 100000 times

#include <stdio.h>
#include <string.h>

char testnum[17];

void init_num(void) {
    strcpy(testnum, "4532015112830366");
}

int digit_at(int i) {
    return testnum[i] - '0';
}

int luhn_check(void) {
    int sum = 0;
    for (int i = 0; i < 16; i++) {
        int d = digit_at(15 - i);
        if (i % 2) {
            d *= 2;
            if (d > 9) d -= 9;
        }
        sum += d;
    }
    return sum % 10 == 0;
}

int main(void) {
    init_num();
    int result = 0;
    for (int i = 0; i < 100000; i++) {
        if (luhn_check()) {
            result += 2;
        } else {
            result++;
        }
    }
    printf("%d\n", result);
    return 0;
}

File: compiler/bench/luhn.fs
\ expected: 200000
\ Luhn checksum benchmark - validate credit card numbers, run 100000 times

create testnum 16 allot

: init-num ( -- )
  s" 4532015112830366" drop testnum 16 cmove ;

: digit-at ( i -- d )
  testnum + c@ [char] 0 - ;

: luhn-check ( -- valid )
  0
  16 0 do
    15 i - digit-at
    i 2 mod if
      2* dup 9 > if 9 - then
    then
    +
  loop
  10 mod 0= ;

: main ( -- )
  init-num
  0
  100000 0 do
    luhn-check if 2 + else 1+ then
  loop
  . cr ;
